cty/function/stdlib/collection.go

```diff
@@ -649,28 +670,33 @@ var LookupFunc = function.New(&function.Spec{
 		}
 	},
 	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+		// leave default value marked
 		defaultVal := args[2]
 
-		mapVar := args[0]
-		lookupKey := args[1].AsString()
+		// unmark collection, retain marks to reapply later
+		mapVar, marks := args[0].Unmark()
+
+		// disregard marks on the key
+		keyVal, _ := args[1].Unmark()
```

Comment:
My intuition here would be that we'd want to merge any marks on the key string into the result's marks so that we don't "lose" them. For example, in Terraform's case if I try to use a sensitive key to look into a non-sensitive map then the result it ends up selecting could imply what the sensitive key was. Does that make sense, or am I missing a subtlety here? :thinking: 

It's a shame that this function ends up having to conditionally use either `GetAttr` or `Index` below because `Index` already handles its argument being marked because an index is a `cty.Value` itself, but attributes are always just normal Go strings and so it ends up being the caller's responsibility in that case. I guess there's a possible restructure here wkere we only do the `Unmark` and `AsString` inside the `mapVar.Type().IsObjectType()` branch below, and just let  the marked `keyVal` pass directly into the `mapVar.Index` call, but I guess that's a tradeoff between making the map case definitely match how direct indexing would work (because it uses the same function) vs. potentially causing the object and map cases to diverge due to being implemented differently. So I think I'd lean towards leaving the structure as-is for now but to add the key marks into `marks` to be represented in the return value, unless we decide against doing that latter part after discussion.

Reply:
> For example, in Terraform's case if I try to use a sensitive key to look into a non-sensitive map then the result it ends up selecting could imply what the sensitive key was

This is true, but it's also not my understanding of the kind of threats we're trying to model in Terraform. We're only trying to prevent accidental disclosure in UI, and since the key is not visible in the output of `Lookup`, its marks are irrelevant in that case.

I can imagine that a more conservative approach would make sense for other uses of marks, but I can't think of any examples where that makes sense. My rule of thumb in these cases has been to propagate marks on values which may be present in the output (e.g. string transformations, object subsets) but not on values which are intermediate (e.g. list indices, map keys).

Reply:
Just noting I'll happily leave the decision on this up to you—I've pushed a fixup commit 
(b535a2a) which I think does what you originally suggested here. If you're okay with doing the cleanup, please either squash that down or drop it before merge, depending on which direction you want to go.